go/consensus/tendermint/abci/state.go

```diff
@@ -114,101 +62,85 @@ type applicationState struct { // nolint: maligned
-	}
-	c.ctx = context.WithValue(s.ctx, contextKey{}, c)
-
+	var blockCtx *api.BlockContext
+	var state mkvs.Tree
 	switch mode {
-	case ContextInitChain:
-		c.state = s.deliverTxTree
-	case ContextCheckTx:
-		c.state = s.checkTxTree
-	case ContextDeliverTx, ContextBeginBlock, ContextEndBlock:
-		c.state = s.deliverTxTree
-		c.blockCtx = s.blockCtx
-	case ContextSimulateTx:
+	case api.ContextInitChain:
+		state = s.deliverTxTree
+	case api.ContextCheckTx:
+		state = s.checkTxTree
+	case api.ContextDeliverTx, api.ContextBeginBlock, api.ContextEndBlock:
+		state = s.deliverTxTree
+		blockCtx = s.blockCtx
+	case api.ContextSimulateTx:
 		// Since simulation is running in parallel to any changes to the database, we make sure
 		// to create a separate in-memory tree at the given block height.
-		c.state = iavl.NewMutableTree(s.db, 128)
-		// NOTE: This requires a specific implementation of `LoadVersion` which doesn't rely
-		//       on cached metadata. Such an implementation is provided in our fork of IAVL.
-		if _, err := c.state.LoadVersion(c.blockHeight); err != nil {
-			panic(fmt.Errorf("context: failed to load state at height %d: %w", c.blockHeight, err))
-		}
-		c.currentTime = s.blockTime
+		state = mkvs.NewWithRoot(nil, s.storage.NodeDB(), s.stateRoot, mkvs.WithoutWriteLog())
+		now = s.blockTime
 	default:
 		panic(fmt.Errorf("context: invalid mode: %s (%d)", mode, mode))
 	}
 
-	return c
+	return api.NewContext(
+		s.ctx,
+		mode,
+		now,
+		api.NewNopGasAccountant(),
+		s,
+		state,
+		int64(s.stateRoot.Round),
+		blockCtx,
+	)
+}
+
+func (s *applicationState) Storage() storage.LocalBackend {
+	return s.storage
 }
 
-// BlockHeight returns the last committed block height.
 func (s *applicationState) BlockHeight() int64 {
 	s.blockLock.RLock()
 	defer s.blockLock.RUnlock()
 
-	return s.blockHeight
+	return int64(s.stateRoot.Round)
 }
 
-// BlockHash returns the last committed block hash.
 func (s *applicationState) BlockHash() []byte {
 	s.blockLock.RLock()
 	defer s.blockLock.RUnlock()
 
-	return append([]byte{}, s.blockHash...)
+	if s.stateRoot.Round == 0 {
+		// Tendermint expects a nil hash when there is no state otherwise it will panic.
+		return nil
+	}
+	return s.stateRoot.Hash[:]
 }
 
-// BlockContext returns the current block context which can be used
-// to store intermediate per-block results.
-//
-// This method must only be called from BeginBlock/DeliverTx/EndBlock
-// and calls from anywhere else will cause races.
-func (s *applicationState) BlockContext() *BlockContext {
-	return s.blockCtx
-}
+func (s *applicationState) ConsensusParameters() *consensusGenesis.Parameters {
+	s.blockLock.RLock()
+	defer s.blockLock.RUnlock()
 
-// DeliverTxTree returns the versioned tree to be used by queries
-// to view comitted data, and transactions to build the next version.
-func (s *applicationState) DeliverTxTree() *iavl.MutableTree {
-	return s.deliverTxTree
+	return s.blockParams
```

Comment:
we unlock and return a pointer. so the rest of the codebase has to be respectful of the deep immutability